### PR TITLE
add RPM variant to libabigail

### PR DESCRIPTION
**Pull request by woodard**
Add a variant to libabigail package which builds the RPM tools.  There
were two python dependencies missing py-xdg and py-koji and so I also
created packages for them.

Signed-off-by: Ben Woodard <woodard@redhat.com>

### CODE REVIEW

**Comment by woodard** [comment]
This is more of an RFC at the moment. It hasn't passed my local testing. For some reason the python include path is not getting set properly or something and so the tools are not able to find their dependencies.

```
$ fedabipkgdiff --self-compare -a --from fc36 libabigail
Traceback (most recent call last):
  File "/home/ben/Shared/Work/spack/opt/spack/linux-fedora36-skylake/gcc-12.1.1/libabigail-master-zgh7obpagd73icmgnrq3p27fadjyhzgm/bin/fedabipkgdiff", line 27, in <module>
    import xdg.BaseDirectory
ModuleNotFoundError: No module named 'xdg.BaseDirectory'
```

**Comment by woodard** [comment]
Another thing that I noticed is doing `spack load libabigail` does seem to put py-xdg from the spack build into the PYTHONPATH. This makes the error message that much more cryptic.

File: var/spack/repos/builtin/packages/py-koji/package.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+class PyKoji(PythonPackage):
+    """Python library to access the Koji system for building and tracking RPMS."""
+
+    pypi = "koji/koji-1.29.0.tar.gz"
+
+    version('1.29.0', sha256='7db1ea4f68dd94d556c612e40bbfc2268164bfebc4a4cd08b63971bddaa686bc')
+
+    extends('python', ignore=r'bin/pytest')
+    depends_on('py-python-dateutil', type=('build', 'run'))
+    depends_on('py-requests', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-koji/package.py:18]: Missing deps on requests-gssapi and six

File: var/spack/repos/builtin/packages/py-koji/package.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    depends_on('python@3.6:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-koji/package.py]:
```suggestion
    depends_on('python@2.7:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-pyxdg/package.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+class PyPyxdg(PythonPackage):
+    """Python library to access freedesktop.org standards."""
+
+    pypi = "pyxdg/pyxdg-0.28.tar.gz"
+
+    version('0.28', sha256='3267bb3074e934df202af2ee0868575484108581e6f3cb006af1da35395e88b4')
+
+    extends('python', ignore=r'bin/pytest')
+    depends_on('python@3.6:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyxdg/package.py]: Where do you see a 3.6+ requirement? I think you can remove this line

File: var/spack/repos/builtin/packages/py-pyxdg/package.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#    depends_on('py-poetry-core@1.0.0:', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyxdg/package.py]: This line can be removed

File: var/spack/repos/builtin/packages/py-koji/package.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class PyKoji(PythonPackage):
+    """Python library to access the Koji system for building and tracking RPMS."""
+
+    pypi = "koji/koji-1.29.0.tar.gz"
+
+    version('1.29.0', sha256='7db1ea4f68dd94d556c612e40bbfc2268164bfebc4a4cd08b63971bddaa686bc')
+
+    extends('python', ignore=r'bin/pytest')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-koji/package.py:16]: Is this actually needed? I think this only matters when using `spack activate` which we are in the process of removing.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Will try building locally to see why you're getting that error message.

File: var/spack/repos/builtin/packages/py-koji/package.py (hunk 5 above)
Comment (woodard) [reply var/spack/repos/builtin/packages/py-koji/package.py:16]: I don't follow you on this one.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-koji/package.py:16]: I don't think `ignore='bin/pytest'` should be needed. Why did you add it?

**Comment by adamjstewart** [comment]
Tried building locally but rpm fails to build with Apple Clang 12.0.0:
```
lrexlib.c:35:27: error: implicit declaration of function 'lua_strlen' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  pr->re_endp = pattern + lua_strlen(L, 1);
                          ^
1 error generated.
make[2]: *** [libluaext_la-lrexlib.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/libluaext_la-lposix.Tpo .deps/libluaext_la-lposix.Plo
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

**Comment by adamjstewart** [comment]
To debug this I would:

1. Uninstall and reinstall these 4 packages, maybe you have an old copy lying around
2. Run `spack load libabigail` again
3. Check `echo $PYTHONPATH` and make sure it contains `py-pyxdg`
4. Try the same import yourself in Python
5. Make sure you don't have an `xdg` directory in your current directory that is masking the import

FWIW, this worked for me:
```console
$ spack load py-pyxdg
$ python
Python 3.9.13 (main, Jun 21 2022, 22:34:57) 
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import xdg.BaseDirectory
>>>
```

File: var/spack/repos/builtin/packages/py-koji/package.py (hunk 1 above)
Comment (woodard) [reply var/spack/repos/builtin/packages/py-koji/package.py:18]: I do not see where requests-gssapi is required.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-koji/package.py:18]: In the `setup.py` file. If you're unfamiliar with Python's build system, https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html is a helpful read. This package uses setuptools so you'll want to pay close attention to the setuptools section.

File: var/spack/repos/builtin/packages/libabigail/package.py [hunk 6]
```diff
@@ -12,12 +12,14 @@ class Libabigail(AutotoolsPackage):
...
     url      = "https://mirrors.kernel.org/sourceware/libabigail/libabigail-2.0.tar.gz"
     git      = "https://sourceware.org/git/libabigail.git"
+    maintainers = ['woodard']
 
     version('master', branch='master')
     version('2.0', sha256='3704ae97a56bf076ca08fb5dea6b21db998fbbf14c4f9de12824b78db53b6fda')
     version('1.8', sha256='1cbf260b894ccafc61b2673ba30c020c3f67dbba9dfa88dca3935dff661d665c')
 
     variant('docs', default=False, description='build documentation')
+    variant('rpm', default=False, description='build RPM tools')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/libabigail/package.py:22]: Would it make sense to have this variant only on linux::
```suggestion
    variant('rpm', default=False, when='platform=linux', description='build RPM tools')
```
?

**Comment by alalazo** [comment]
@woodard

**Comment by alecbcs** [comment]
Going to go head and close this PR due to a lack of response. @woodard happy to reopen if you still need this variant added.

**Pull request closed without merging**